scenario/clickactionbutton.go

```diff
@@ -0,0 +1,643 @@
+		return errors.Errorf("Empty object ID")
+	}
+	return nil
+}
+
+// Execute button-actions contained in sense action-button
+func (settings ClickActionButtonSettings) Execute(sessionState *session.State, actionState *action.State, connectionSettings *connection.ConnectionSettings, label string, reset func()) {
+	if sessionState.Connection == nil || sessionState.Connection.Sense() == nil {
+		actionState.AddErrors(errors.New("Not connected to a Sense environment"))
+		return
+	}
+
+	if sessionState.Connection.Sense().CurrentApp == nil {
+		actionState.AddErrors(errors.New("Not connected to a Sense app"))
+		return
+	}
+
+	// retrieve action-button-object
+	objectID := sessionState.IDMap.Get(settings.ID)
+	obj, err := sessionState.Connection.Sense().Objects.GetObjectByID(objectID)
+	if err != nil {
+		actionState.AddErrors(errors.Wrapf(err, "Failed getting object<%s> from object list", obj.ID))
+		return
+	}
+
+	// retrieve button-actions
+	buttonActions, navigationAction, err := buttonActions(sessionState, actionState, obj)
+	if err != nil {
+		actionState.AddErrors(errors.Wrapf(err, "Failed to get button actions"))
+		return
+	}
+
+	// run button-actions
+	for _, buttonAction := range buttonActions {
+		err := buttonAction.execute(sessionState, actionState)
+		if err != nil {
+			actionState.AddErrors(errors.Wrapf(err, "Buttonaction type<%s> label<%s> cid<%s> failed",
+				buttonAction.ActionType, buttonAction.ActionLabel, buttonAction.CID))
+		}
+	}
+
+	err = navigationAction.execute(sessionState, actionState, connectionSettings, label)
+	if err != nil {
+		actionState.AddErrors(errors.Wrapf(err, "Button-navigationaction<%s> failed", navigationAction.Action))
+	}
+
+	sessionState.Wait(actionState)
+}
+
+// buttonAction returns button actions and navigation action for obj
+func buttonActions(sessionState *session.State, actionState *action.State, obj *enigmahandlers.Object) ([]buttonAction, *buttonNavigationAction, error) {
+	// TODO(atluq): Add buttonActions to enigma.GenericObjectProperties such
+	// that obj.Properties() could be used, instead of
+	// obj.EnigmaObject.GetpropertiesRaw()
+
+	switch t := obj.EnigmaObject.(type) {
+	case *enigma.GenericObject:
+		genericObj := obj.EnigmaObject.(*enigma.GenericObject)
+
+		propsRaw, err := sessionState.SendRequestRaw(actionState, genericObj.GetPropertiesRaw)
+		if err != nil {
+			return nil, nil, errors.Wrapf(err, "Properties request failed for object<%s>", obj.ID)
+		}
+
+		// parse button actions
+		actionsRaw, err := senseobjdef.NewDataPath("actions").Lookup(propsRaw)
+		if err != nil {
+			return nil, nil, errors.Wrapf(err, `No "actions"-property exist for object<%s>`, obj.ID)
+		}
+
+		actions := []buttonAction{}
+		err = json.Unmarshal(actionsRaw, &actions)
+		if err != nil {
+			return nil, nil, errors.Wrapf(err, "Failed to unmarshal button actions")
+		}
+
+		// parse navigation action associated with button
+		navigationRaw, err := senseobjdef.NewDataPath("navigation").Lookup(propsRaw)
```

Comment:
Do both actions and navigation exist for all buttons? So there's no buttons with exclusively one or the other?

Reply:
There exist buttons with one or the other, but the fields are never omitted in messages. There is a "none" navigation action and an empty list for the other actions.